File: examples/python/lockr_example.py
"""
Lockr Python client example using Ed25519 challenge-response auth.
Requires: pip install httpx cryptography
"""
import httpx
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat
import binascii


class LockrClient:
    def __init__(self, addr: str, key_path: str, ca_path: str | None = None):
        self.addr = addr.rstrip("/")
        with open(key_path) as f:
            key_hex = f.read().strip()
        raw = bytes.fromhex(key_hex)
        # Ed25519 private key is 64 bytes: first 32 = seed, last 32 = public key
        self._privkey = Ed25519PrivateKey.from_private_bytes(raw[:32])
        self._token: str | None = None
        verify = ca_path if ca_path else False
        self._http = httpx.Client(base_url=self.addr, verify=verify, timeout=15)

    def authenticate(self, service: str) -> None:
        r = self._http.post("/v1/auth/challenge", json={"service": service})
        r.raise_for_status()
        challenge_hex = r.json()["data"]["challenge"]
        challenge = bytes.fromhex(challenge_hex)
        sig = self._privkey.sign(challenge)
        r2 = self._http.post("/v1/auth/verify", json={
            "challenge": challenge_hex,
            "signature": sig.hex(),
        })
        r2.raise_for_status()
        self._token = r2.json()["data"]["token"]
        self._http.headers.update({"Authorization": f"Bearer {self._token}"})

    def kv_get(self, path: str, version: int = 0) -> dict:
        url = f"/v1/secrets/kv/{path}"
        if version:
            url += f"?version={version}"
        r = self._http.get(url)
        r.raise_for_status()
        return r.json()["data"]["value"]

    def kv_set(self, path: str, value: dict) -> None:
        r = self._http.put(f"/v1/secrets/kv/{path}", json=value)
        r.raise_for_status()

    def transit_encrypt(self, key_name: str, plaintext: str) -> str:
        r = self._http.post(f"/v1/secrets/transit/{key_name}/encrypt",
                            json={"plaintext": plaintext})
        r.raise_for_status()
        return r.json()["data"]["ciphertext"]

    def transit_decrypt(self, key_name: str, ciphertext: str) -> str:
        r = self._http.post(f"/v1/secrets/transit/{key_name}/decrypt",
                            json={"ciphertext": ciphertext})
        r.raise_for_status()
        return r.json()["data"]["plaintext"]


if __name__ == "__main__":
    vault = LockrClient(
        addr="https://lockr:8300",
        key_path="./certs/api-server.key",
        ca_path="./certs/ca.crt",
    )
    vault.authenticate("api-server")

    stripe = vault.kv_get("secrets/prod/stripe_key")
    print("stripe key:", stripe["key"])

    ct = vault.transit_encrypt("payments-key", "card:4111111111111111")
    print("encrypted:", ct)

    pt = vault.transit_decrypt("payments-key", ct)
    print("decrypted:", pt)

File: examples/nodejs/lockr_example.js
/**
 * Lockr Node.js client example using Ed25519 challenge-response auth.
 * Requires: npm install node-fetch
 * Node 20+ has built-in crypto.sign for Ed25519.
 */
const { createPrivateKey, sign } = require("crypto");
const { readFileSync } = require("fs");

class LockrClient {
  constructor({ addr, keyPath, caPath }) {
    this.addr = addr.replace(/\/$/, "");
    const keyHex = readFileSync(keyPath, "utf8").trim();
    const rawKey = Buffer.from(keyHex, "hex");
    // PKCS8 wrapper for Ed25519 (seed = first 32 bytes)
    const pkcs8Prefix = Buffer.from(
      "302e020100300506032b657004220420",
      "hex"
    );
    this._privkeyDer = Buffer.concat([pkcs8Prefix, rawKey.slice(0, 32)]);
    this._token = null;
    this._caPath = caPath;
  }

  async _fetch(method, path, body) {
    const https = require("https");
    const agent = this._caPath
      ? new https.Agent({ ca: readFileSync(this._caPath) })
      : new https.Agent({ rejectUnauthorized: false });

    const { default: fetch } = await import("node-fetch");
    const res = await fetch(this.addr + path, {
      method,
      headers: {
        "Content-Type": "application/json",
        ...(this._token ? { Authorization: `Bearer ${this._token}` } : {}),
      },
      body: body ? JSON.stringify(body) : undefined,
      agent,
    });
    return res.json();
  }

  async authenticate(service) {
    const r1 = await this._fetch("POST", "/v1/auth/challenge", { service });
    const challengeHex = r1.data.challenge;
    const challenge = Buffer.from(challengeHex, "hex");

    const privKey = createPrivateKey({ key: this._privkeyDer, format: "der", type: "pkcs8" });
    const sig = sign(null, challenge, privKey);

    const r2 = await this._fetch("POST", "/v1/auth/verify", {
      challenge: challengeHex,
      signature: sig.toString("hex"),
    });
    this._token = r2.data.token;
  }

  async kvGet(path, version = 0) {
    const url = version ? `${path}?version=${version}` : path;
    const r = await this._fetch("GET", `/v1/secrets/kv/${url}`);
    return r.data?.value ?? r.data;
  }

  async kvSet(path, value) {
    return this._fetch("PUT", `/v1/secrets/kv/${path}`, value);
  }

  async transitEncrypt(keyName, plaintext) {
    const r = await this._fetch("POST", `/v1/secrets/transit/${keyName}/encrypt`, { plaintext });
    return r.data.ciphertext;
  }

  async transitDecrypt(keyName, ciphertext) {
    const r = await this._fetch("POST", `/v1/secrets/transit/${keyName}/decrypt`, { ciphertext });
    return r.data.plaintext;
  }
}

(async () => {
  const vault = new LockrClient({
    addr: "https://lockr:8300",
    keyPath: "./certs/api-server.key",
    caPath: "./certs/ca.crt",
  });

  await vault.authenticate("api-server");

  const stripe = await vault.kvGet("secrets/prod/stripe_key");
  console.log("stripe key:", stripe.key);

  const ct = await vault.transitEncrypt("payments-key", "card:4111111111111111");
  console.log("encrypted:", ct);

  const pt = await vault.transitDecrypt("payments-key", ct);
  console.log("decrypted:", pt);
})();
